src/ring_theory/graded_algebra/basic.lean

```diff
@@ -140,3 +139,45 @@ begin
 end
 
 end graded_algebra
+
+section homogeneous_element
+open_locale pointwise
+
+variables {ι R A : Type*}
+variables [comm_semiring R] [ring A] [algebra R A]
+variables (𝒜 : ι → submodule R A)
+
+/-- An element `a : A` is said to be homogeneous if there is some
+`i : ι` such that `a ∈ 𝒜 i`-/
+def is_homogeneous (a : A) : Prop := ∃ i, a ∈ 𝒜 i
+
+/--We collect all homogeneneous elements into a subtype `homogeneous_element` -/
+def homogeneous_element : Type* := {a // is_homogeneous 𝒜 a}
+
+variables [decidable_eq ι] [add_comm_monoid ι] [graded_algebra 𝒜]
+
+instance homogeneous_element.has_mul :
+  has_mul (homogeneous_element 𝒜) :=
+{ mul := λ ⟨x, h⟩ ⟨y, h'⟩, ⟨x * y, begin
+  rcases h with ⟨i, hi⟩,
+  rcases h' with ⟨j, hj⟩,
+  use (i + j), exact set_like.has_graded_mul.mul_mem hi hj,
+end ⟩ }
+
+/--Converting homogeneous element to `A`-/
+def homogeneous_element.coe :
+  mul_hom (homogeneous_element 𝒜) A :=
+{ to_fun := λ x, x.1,
+  map_mul' := λ x y, by tidy }
+
+instance lift_homogeneous_element : has_lift (homogeneous_element 𝒜) A :=
+{ lift := homogeneous_element.coe 𝒜 }
+
+instance lift_homogeneous_set :
+  has_lift (set (homogeneous_element 𝒜)) (set A) :=
+{ lift := λ S, (homogeneous_element.coe 𝒜) '' S }
+
+instance homogeneous_element_inhabited : inhabited (homogeneous_element 𝒜) :=
+⟨⟨1, ⟨0, set_like.has_graded_one.one_mem⟩⟩⟩
```

Comment:
I suspect you don't really want any of this; if you jump straight to producing `homogenous_submonoid` as I do above, then you get all this for free via the submonoid API, using `homogenous_submonoid 𝒜` instead of `homogeneous_element 𝒜`.